pinot-druid/src/main/java/org/apache/pinot/druid/data/readers/DruidSegmentRecordReader.java

```diff
@@ -0,0 +1,219 @@
+    _columnNames.addAll(_pinotSchema.getColumnNames());
+
+    ColumnConfig config = new DruidProcessingConfig() {
+      @Override
+      public String getFormatString() {
+        return "processing-%s";
+      }
+
+      @Override
+      public int intermediateComputeSizeBytes() {
+        return 100 * 1024 * 1024;
+      }
+
+      @Override
+      public int getNumThreads() {
+        return 1;
+      }
+
+      @Override
+      public int columnCacheSizeBytes() {
+        return 25 * 1024 * 1024;
+      }
+    };
+
+    ObjectMapper mapper = new DefaultObjectMapper();
+    final IndexIO indexIO = new IndexIO(mapper, config);
+    _index = indexIO.loadIndex(new File(indexPath));
+    _adapter = new QueryableIndexStorageAdapter(_index);
+
+    // A Sequence "represents an iterable sequence of elements. Unlike normal Iterators however, it doesn't expose
+    // a way for you to extract values from it, instead you provide it with a worker (an Accumulator) and that defines
+    // what happens with the data."
+    final Sequence<Cursor> cursors = _adapter.makeCursors(
+        Filters.toFilter(null),
+        _index.getDataInterval().withChronology(ISOChronology.getInstanceUTC()),
+        VirtualColumns.EMPTY,
+        Granularities.ALL,
+        false,
+        null
+    );
+
+    List<Cursor> cursorList = new ArrayList<>();
+    final Sequence<Object> sequence = Sequences.map(cursors, new Function<Cursor, Object>() {
+      @Override
+      public Object apply(@Nullable Cursor cursor) {
+        cursorList.add(cursor);
+        return null;
+      }
+    });
+    sequence.accumulate(null, (accumulated, in) -> null);
+
+    // There should only be one single Cursor for every segment, so there should only be one Cursor in the cursorList
+    Preconditions.checkArgument(cursorList.size() == 1, "msg");
+    _cursor = cursorList.get(0);
+
+    validateColumns();
+
+    ColumnSelectorFactory columnSelectorFactory = _cursor.getColumnSelectorFactory();
+    _selectors = _columnNames
+        .stream()
+        .map(columnSelectorFactory::makeColumnValueSelector)
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public void init(SegmentGeneratorConfig segmentGeneratorConfig)
+      throws IOException {
+    init(segmentGeneratorConfig.getInputFilePath(), segmentGeneratorConfig.getSchema());
+  }
+
+  @Override
+  public boolean hasNext() {
+    return !_cursor.isDone();
+  }
+
+  @Override
+  public GenericRow next() throws IOException {
+    return next(new GenericRow());
+  }
+
+  @Override
+  public GenericRow next(GenericRow reuse) {
+    // Only the columns whose names are in the Pinot schema will get processed
+    for (int i = 0; i < _columnNames.size(); i++) {
+      final String columnName = _columnNames.get(i);
+      BaseObjectColumnValueSelector selector = _selectors.get(i);
+      // If the column does not exist in the segment file, skip it.
+      if (selector != null) {
+        final Object value = selector.getObject();
```

Comment:
Have we checked what types of object returned by Druid for all possible Pinot columns types? Can you add this mapping information to the design document?

If we are sure that for all 

e.g. For long column, `selector.getObject()` returns `Long` object in Druid, and Pinot also uses `Long` obejct for `GenericRow`

For multi-value long column, `select.getObject()` returns `???` object in Druid, and Pinot uses `Long[]` Long array for `GenericRow` etc.